src/main/scala/firrtl/passes/VerilogMemDelays.scala

```diff
@@ -0,0 +1,171 @@
+     software without specific prior written permission.
+IN NO EVENT SHALL REGENTS BE LIABLE TO ANY PARTY FOR DIRECT, INDIRECT,
+SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES, INCLUDING LOST PROFITS,
+ARISING OUT OF THE USE OF THIS SOFTWARE AND ITS DOCUMENTATION, EVEN IF
+REGENTS HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+REGENTS SPECIFICALLY DISCLAIMS ANY WARRANTIES, INCLUDING, BUT NOT
+LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+A PARTICULAR PURPOSE. THE SOFTWARE AND ACCOMPANYING DOCUMENTATION, IF
+ANY, PROVIDED HEREUNDER IS PROVIDED "AS IS". REGENTS HAS NO OBLIGATION
+TO PROVIDE MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS, OR
+MODIFICATIONS.
+*/
+
+package firrtl.passes
+
+import firrtl._
+import firrtl.ir._
+import firrtl.Utils._
+import firrtl.Mappers._
+import firrtl.PrimOps._
+
+/** This pass generates delay reigsters for memories for verilog */
+object VerilogMemDelays extends Pass {
+  def name = "Verilog Memory Delays"
+  val ug = UNKNOWNGENDER
+  type Netlist = collection.mutable.HashMap[String, Expression]
+  implicit def expToString(e: Expression) = e.serialize
+  private def NOT(e: Expression) = DoPrim(Not, Seq(e), Nil, BoolType)
+  private def AND(e1: Expression, e2: Expression) = DoPrim(And, Seq(e1, e2), Nil, BoolType)
+
+  def buildNetlist(netlist: Netlist)(s: Statement): Statement = {
+    s match {
+      case Connect(_, loc, expr) => kind(loc) match {
+        case MemKind(_) => netlist(loc) = expr
+        case _ =>
+      }
+      case _ =>
+    }
+    s map buildNetlist(netlist)
+  }
+
+  def memDelayStmt(netlist: Netlist, namespace: Namespace, repl: Netlist)(
+      s: Statement): Statement = s map memDelayStmt(netlist, namespace, repl) match {
+    case s: DefMemory =>
+      import MemPortUtils._
+      val ports = (s.readers ++ s.writers).toSet
+      def newPortName(rw: String, p: String) = (for {
+        idx <- Stream from 0
+        newName = s"${rw}_${p}_${idx}"
+        if !ports(newName)
+      } yield newName).head
+      val rwMap = s.readwriters map (rw =>
+        rw -> (newPortName(rw, "r"), newPortName(rw, "w"))) toMap
+      // 1. readwrite ports are split into read & write ports
+      // 2. memories are transformed into combinational
+      //    because latency pipes are added for longer latencies
+      val mem = s copy (
+        readers = (s.readers ++ (s.readwriters map (rw => rwMap(rw)._1))),
+        writers = (s.writers ++ (s.readwriters map (rw => rwMap(rw)._2))),
+        readwriters = Nil, readLatency = 0, writeLatency = 1)
+      def pipe(e: Expression, n: Int, clk: Expression, cond: Expression): (Expression, Seq[Statement]) = {
+        val node = DefNode(NoInfo, namespace.newTemp, netlist(e))
+        val wref = WRef(node.name, e.tpe, NodeKind(), MALE)
+        ((0 until n) foldLeft (wref, Seq[Statement](node))){case ((ex, stmts), i) =>
+          val name = namespace newName s"${LowerTypes.loweredName(e)}_pipe_$i"
```

Comment:
You *could* (not saying you *should*) let the namespace itself determine how best to disambiguate names. It will do _$i for you (although the first one will not have _$i at the end so doing it manually for the sake of consistency makes sense)